zookeeper/src/test/java/com/linecorp/armeria/client/endpoint/zookeeper/ZookeeperEndpointGroupTest.java

```diff
@@ -103,6 +116,30 @@ public void testGetEndpointGroup() {
         assertEquals(initializedEndpointGroupList, zookeeperEndpointGroup.endpoints());
     }
 
+    @Test
+    public void testConnectionRecovery() {
+        ZooKeeper zkHandler1 = zookeeperEndpointGroup.getZkHandler();
+        CountDownLatch latch = new CountDownLatch(1);
+        ZooKeeper zkHandler2;
+        try {
+            //create a new handler with the same sessionId and password
+            zkHandler2 = new ZooKeeper(zkInstance.connectString().get(), sessionTimeout, event -> {
+                if (event.getState() == KeeperState.SyncConnected) {
+                    latch.countDown();
+                }
+            }, zkHandler1.getSessionId(), zkHandler1.getSessionPasswd());
+            latch.await();
+            //once connected, close the new handler to cause the original handler session expire
+            zkHandler2.close();
+            //wait  for the session to be expired
+            Thread.sleep(retryWait);
```

Comment:
If we use `BlockingQueue` for storing the state, we don't need to sleep for 5 seconds here.